File: bowling-game/bowling-game-cs/BowlingGame/01/Game.cs
namespace BowlingGame._01;

public class Game {
    readonly int[] rolls = new int[21];
    int currentRoll;

    public void Roll(int pins) {
        rolls[currentRoll++] = pins;
    }

    public int Score() {
        var score = 0;
        var firstInFrame = 0;
        for (var frame = 0; frame < 10; frame++)
            if (IsStrike(firstInFrame)) {
                score += 10 + NextTwoBallsForStrike(firstInFrame);
                firstInFrame++;
            }
            else if (IsSpare(firstInFrame)) {
                score += 10 + NextBallForSpare(firstInFrame);
                firstInFrame += 2;
            }
            else {
                score += TwoBallsInFrame(firstInFrame);
                firstInFrame += 2;
            }

        return score;
    }

    int NextTwoBallsForStrike(int firstInFrame) {
        return rolls[firstInFrame + 1] + rolls[firstInFrame + 2];
    }

    int NextBallForSpare(int firstInFrame) {
        return rolls[firstInFrame + 2];
    }

    int TwoBallsInFrame(int firstInFrame) {
        return rolls[firstInFrame] + rolls[firstInFrame + 1];
    }

    bool IsStrike(int firstInFrame) {
        return rolls[firstInFrame] == 10;
    }

    bool IsSpare(int firstInFrame) {
        return rolls[firstInFrame] + rolls[firstInFrame + 1] == 10;
    }
}

File: bowling-game/bowling-game-py/07/Game.py
class Game:
    def __init__(self):
        self.rolls = []

    def roll(self, pins) -> None:
        self.rolls.append(pins)

    def score(self) -> int:
        score = 0
        first_in_frame = 0
        for frame in range(10):
            if self.is_strike(first_in_frame):
                score += 10 + self.two_balls_in_frame(first_in_frame + 1)
                first_in_frame += 1
            elif self.is_spare(first_in_frame):
                score += 10 + self.next_ball_for_spare(first_in_frame)
                first_in_frame += 2
            else:
                score += self.two_balls_in_frame(first_in_frame)
                first_in_frame += 2
        return score

    def is_strike(self, first_in_frame) -> bool:
        return self.rolls[first_in_frame] == 10

    def is_spare(self, first_in_frame) -> bool:
        return self.two_balls_in_frame(first_in_frame) == 10

    def next_ball_for_spare(self, first_in_frame) -> int:
        return self.rolls[first_in_frame + 2]

    def two_balls_in_frame(self, first_in_frame) -> int:
        return self.rolls[first_in_frame] + self.rolls[first_in_frame + 1]